File: mobile/android/app/src/main/java/com/findly/android/queue/worker/LocationSyncRunner.kt
package com.findly.android.queue.worker

import com.findly.android.location.FixCaptureCoordinator
import com.findly.android.location.settings.DeviceSettingsCoordinator
import com.findly.android.location.settings.GeofenceConfigSyncCoordinator
import com.findly.android.queue.FixSource
import com.findly.android.queue.GeofenceEventSyncCoordinator
import com.findly.android.queue.GeofenceEventSyncOutcome
import com.findly.android.queue.LocationSyncCoordinator
import com.findly.android.queue.SyncOutcome
import java.time.LocalDate

/** [LocationSyncRunner.runOnce]'s result — thin enough for both `LocationSyncWorker` (maps
 * straight to `Result.success()`/`Result.retry()`) and the foreground-service loop (applies
 * [BackoffPolicy] before its next iteration on [Retry]) to consume directly. */
sealed class RunResult {
    data object Success : RunResult()
    data object Retry : RunResult()
}

/**
 * One periodic-sync cycle (specs/009-device-runtime.md §3.1: "the worker captures a fix and
 * queues it even offline; the upload leg tolerates failure and retries") — deliberately **pure
 * Kotlin, zero `android.*` imports**, so this is where the actual "what does a sync run do" logic
 * lives and is unit-tested; `LocationSyncWorker` (WorkManager) and the §3.2 foreground service are
 * both thin callers of the exact same [runOnce].
 *
 * Order per run: (1) the once-per-local-day gate (§3.3) before ever attempting a periodic
 * capture; (2) capture-and-queue one `source: "periodic"` fix (silently skipped by
 * [FixCaptureCoordinator] itself if paused/no permission/debounced, §1.2); (3) drain the fix queue
 * via [syncCoordinator]; (4) drain the geofence-event queue via [geofenceEventSyncCoordinator]
 * (§6.3: "Events are flushed like fixes... on the same cadence") — the natural fit found by
 * piggybacking onto this same per-cycle drain rather than a second independent scheduler. Every
 * successful flush of *either* queue applies its mandatory `deviceSettings` piggyback (§5.1/§7.3)
 * without stopping the run, and re-syncs the geofence config when the piggybacked `geofenceEtag`
 * differs from the cached one (§6.2/§6.3's ETag-mismatch self-heal); only `403 TRACKING_PAUSED`
 * stops early.
 */
class LocationSyncRunner(
    private val currentSyncIntervalMinutes: suspend () -> Int,
    private val lastCaptureDateStore: LastCaptureDateStore,
    private val today: () -> LocalDate,
    private val captureCoordinator: FixCaptureCoordinator,
    private val syncCoordinator: LocationSyncCoordinator,
    private val geofenceEventSyncCoordinator: GeofenceEventSyncCoordinator,
    private val settingsCoordinator: DeviceSettingsCoordinator,
    private val geofenceConfigSyncCoordinator: GeofenceConfigSyncCoordinator,
    private val onReRegisterDevice: suspend () -> Unit,
    private val onSignedOut: suspend () -> Unit,
) {
    suspend fun runOnce(): RunResult {
        maybeCapturePeriodicFix()
        val fixDrainResult = drainFixQueue()
        // A transient failure on the fix queue backs off the whole run (§9) - no point hammering
        // the geofence-event endpoint too in the same cycle; the next run retries both.
        if (fixDrainResult == RunResult.Retry) return RunResult.Retry
        return drainGeofenceEventQueue()
    }

    private suspend fun maybeCapturePeriodicFix() {
        val interval = currentSyncIntervalMinutes()
        val today = today()
        if (OnceDailyGate.shouldSkipCapture(interval, today, lastCaptureDateStore.lastCaptureDate())) return

        val captured = captureCoordinator.captureAndQueue(FixSource.Periodic)
        if (captured != null) lastCaptureDateStore.recordCaptureDate(today)
    }

    private suspend fun drainFixQueue(): RunResult {
        repeat(MAX_BATCHES_PER_RUN) {
            val outcome = syncCoordinator.syncOnce()
            if (outcome is SyncOutcome.NothingToSync) return RunResult.Success

            when (val reaction = SyncOutcomeReactor.reactionFor(outcome)) {
                SyncReaction.Continue -> Unit // more of the queue may remain - loop again
                is SyncReaction.Synced -> applySyncedPiggyback(reaction)
                is SyncReaction.ApplySettings -> {
                    settingsCoordinator.applySettings(reaction.settings)
                    return RunResult.Success
                }
                SyncReaction.ReRegisterDevice -> {
                    onReRegisterDevice()
                    return RunResult.Success
                }
                SyncReaction.SignedOut -> {
                    onSignedOut()
                    return RunResult.Success
                }
                SyncReaction.RetryTransient -> return RunResult.Retry
            }
        }
        // Defensively bounded (§3.1: "MUST... complete in well under 10 minutes") - a
        // pathologically large backlog is picked up again next run rather than looping forever.
        return RunResult.Success
    }

    /** specs/009-device-runtime.md §6.3: "Events are flushed like fixes, batched 1-20 per call". */
    private suspend fun drainGeofenceEventQueue(): RunResult {
        repeat(MAX_BATCHES_PER_RUN) {
            val outcome = geofenceEventSyncCoordinator.syncOnce()
            if (outcome is GeofenceEventSyncOutcome.NothingToSync) return RunResult.Success

            when (val reaction = SyncOutcomeReactor.reactionForGeofenceEvents(outcome)) {
                SyncReaction.Continue -> Unit
                is SyncReaction.Synced -> applySyncedPiggyback(reaction)
                is SyncReaction.ApplySettings -> {
                    settingsCoordinator.applySettings(reaction.settings)
                    return RunResult.Success
                }
                SyncReaction.ReRegisterDevice -> {
                    onReRegisterDevice()
                    return RunResult.Success
                }
                SyncReaction.SignedOut -> {
                    onSignedOut()
                    return RunResult.Success
                }
                SyncReaction.RetryTransient -> return RunResult.Retry
            }
        }
        return RunResult.Success
    }

    /** The shared "apply the mandatory piggyback" step both drain loops' `Synced` branch needs
     * (001-api-contract.md §5.1/§7.3): settings apply unconditionally; the geofence config only
     * re-syncs when the observed etag actually differs from the cached one
     * ([GeofenceConfigSyncCoordinator.syncIfEtagChanged], §6.2/§6.3). */
    private suspend fun applySyncedPiggyback(synced: SyncReaction.Synced) {
        settingsCoordinator.applySettings(synced.deviceSettings)
        geofenceConfigSyncCoordinator.syncIfEtagChanged(synced.geofenceEtag)
    }

    private companion object {
        const val MAX_BATCHES_PER_RUN = 20 // 20 * 100 = 2 000 fixes/run, well above the 1 000 cap
    }
}

File: mobile/ios/FindlyKit/Sources/FindlyKit/Locations/LocationSyncRunner.swift
import Foundation

/// `LocationSyncRunner.runOnce`'s result — thin enough for `SystemBackgroundSyncScheduler`'s
/// `BGAppRefreshTask` launch handler to map straight onto `task.setTaskCompleted(success:)`, and
/// for a future retry-scheduling caller to apply `BackoffPolicy` on `.retry`.
public enum RunResult: Equatable {
    case success
    case retry
}

/// specs/009-device-runtime.md §1 (pipeline) + §3.4 (the 0.8-elapsed-time trigger rule) +
/// §9 (error-handling reactions) — one opportunistic-trigger cycle. All three of iOS's triggers
/// (`BGAppRefreshTask`, significant-location-change already captured via `FixCaptureCoordinator`'s
/// own hint path, foreground) call through this single `runOnce()` — deliberately **pure Swift,
/// zero CoreLocation/BackgroundTasks imports**, so this is where "what does a sync run do" lives
/// and is unit-tested; `LocationRuntimeContainer`'s `BGAppRefreshTask` handler is a thin caller of
/// the exact same `runOnce()`. Mirrors Android's `LocationSyncRunner`, scoped to what I10 owns (no
/// geofence-event queue drain — that's I11's addition, following the same pattern this leaves
/// room for).
///
/// Order per run: (1) maybe capture one `source: .periodic` fix, gated by `SyncTriggerPolicy`'s
/// 0.8 rule (skipped silently if under threshold — the already-queued backlog, if any, still gets
/// drained); (2) drain the fix queue via `syncCoordinator`, applying the mandatory `deviceSettings`
/// piggyback on every `.synced` outcome (specs/001 §5.1, specs/009 §1) and reacting to failures per
/// specs/009 §9's table. Only a `.transientFailure` or `.paused`/`.reRegisterDevice`/`.signedOut`
/// outcome stops the loop early; `.rejected`/`.otherFailure` keep draining (more of the queue may
/// remain after an unmappable rejection un-freezes it for retry).
public final class LocationSyncRunner {
    private let currentSyncIntervalMinutes: () -> Int
    private let lastQueuedFixAtStore: LastQueuedFixAtStoring
    private let now: () -> Date
    private let captureCoordinator: FixCaptureCoordinator
    private let syncCoordinator: LocationSyncCoordinator
    private let settingsApplying: DeviceSettingsApplying
    private let onReRegisterDevice: () async -> Void
    private let onSignedOut: () async -> Void

    /// Defensively bounded (specs/009 §3.1's "MUST complete in well under 10 minutes" applies to
    /// Android's WorkManager path, but the same discipline is cheap insurance here too) - a
    /// pathologically large backlog is picked up again next run rather than looping forever.
    /// 20 * 100 = 2 000 fixes/run, well above the 1 000-fix cap (specs/009 §2).
    private static let maxBatchesPerRun = 20

    public init(
        currentSyncIntervalMinutes: @escaping () -> Int,
        lastQueuedFixAtStore: LastQueuedFixAtStoring,
        now: @escaping () -> Date = Date.init,
        captureCoordinator: FixCaptureCoordinator,
        syncCoordinator: LocationSyncCoordinator,
        settingsApplying: DeviceSettingsApplying,
        onReRegisterDevice: @escaping () async -> Void,
        onSignedOut: @escaping () async -> Void
    ) {
        self.currentSyncIntervalMinutes = currentSyncIntervalMinutes
        self.lastQueuedFixAtStore = lastQueuedFixAtStore
        self.now = now
        self.captureCoordinator = captureCoordinator
        self.syncCoordinator = syncCoordinator
        self.settingsApplying = settingsApplying
        self.onReRegisterDevice = onReRegisterDevice
        self.onSignedOut = onSignedOut
    }

    public func runOnce() async -> RunResult {
        await maybeCapturePeriodicFix()
        return await drainQueue()
    }

    private func maybeCapturePeriodicFix() async {
        let interval = currentSyncIntervalMinutes()
        let shouldCapture = SyncTriggerPolicy.shouldCapture(
            syncIntervalMinutes: interval, lastQueuedFixAt: lastQueuedFixAtStore.lastQueuedFixAt(), now: now()
        )
        guard shouldCapture else { return }
        // FixCaptureCoordinator applies its own §1.2 suppression (paused/permission/debounce) -
        // a nil result here just means nothing was queued, which is fine to treat as "no update
        // to lastQueuedFixAt" (the next trigger will re-evaluate the same 0.8 threshold).
        if await captureCoordinator.captureAndQueue(source: .periodic) != nil {
            lastQueuedFixAtStore.recordQueuedFixAt(now())
        }
    }

    private func drainQueue() async -> RunResult {
        for _ in 0..<Self.maxBatchesPerRun {
            let outcome = await syncCoordinator.syncOnce()
            switch outcome {
            case .nothingToSync:
                return .success

            case .synced(_, _, let deviceSettings, _):
                // 001 §5.1 / 009 §1: applying the piggyback is mandatory on every accepted
                // response, not just a `.paused` nicety - and a successful sync never stops the
                // run, more of the queue may remain.
                await settingsApplying.applySettings(deviceSettings)

            case .paused(let deviceSettings):
                await settingsApplying.applySettings(deviceSettings)
                return .success

            case .reRegisterDevice:
                await onReRegisterDevice()
                return .success

            case .signedOut:
                await onSignedOut()
                return .success

            case .transientFailure:
                return .retry

            case .rejected, .otherFailure:
                continue // the dead/rejected batch is already resolved inside the queue store
            }
        }
        return .success
    }
}
